src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -3503,7 +3503,10 @@ and solve_c (env:Env.env) (problem:problem<comp>) (wl:worklist) : solution =
             | Trivial -> guard
             | NonTrivial f -> U.mk_conj guard f in
           let wl = solve_prob orig (Some <| U.mk_conj guard fml) [] wl in
-          solve env (attempt sub_probs wl) in
+
+          let wl = attempt sub_probs wl in
+          let wl = extend_wl wl g_lift.deferred_to_tac [] in
```

Comment:
Thanks Aymeric, this makes sense to me. We added a new compartment for deferred to tactic constraints in the guard, but forgot to update this part of Rel to carry those constraints forward.

So the guard `g_lift` has 4 fields that need to be propagated to the worklist: `implicits`, `deferred`, `deferred_to_tac`, and the logical formula. Right now we add `implicits` around line 3386, `deferred` around line 3499, and then `deferred_to_tac` around line 3508.

What do you think about extending `extend_wl` to be:

```
let extend_wl (wl:worklist) (deferred:deferred) (defer_to_tac:deferred) (imps:implicits) =
  {wl with  wl_deferred=wl.wl_deferred@deferred;
                wl_deferred_to_tac=wl.wl_deferred_to_tac@(as_wl_deferred wl defer_to_tac);
                wl_implicits=wl.wl_implicits@imps}
```

So then call it at line 3386 with `g.deferred`, `g.deferred_to_tac`, and `g.implicits`, so that we incorporate the guard into the worklist at one place. This will leave only the logical payload in the guard, handling of which can remain as is.

Note that this also means that the line 3499 will no longer add deferred to the sub problems. I think that's fine, since when we try to solve these constraints, they will most probably be deferred again. So adding them to the deferred compartment of the worklist should be fine.

Reply:
That does look like a cleaner way to handle these, I can take care of this. I'll also update the calls in other functions to pass an empty list for deferred, so that the modification will be a noop for other parts of Rel.